### PR TITLE
The Brawlening: Unarmed fighting interactions for shoving, grabbing and nonlethal takedowns (not martial arts)

### PR DESCRIPTION
**Pull request by necromanceranne**
## About The Pull Request

I've tweaked some elements of unarmed fighting to give it additional interactions between the various components, bridging them into a more coherent system and focusing more strongly as tool for disabling opponents nonlethally.

### Shoving

Shoving guarantees that unarmed attacks will land while knocked off-balance (AKA when slowed by a shove).

Being off-balance means that you can be knocked down from a punch if you have taken enough brute and stamina damage combined (at least above 40).

Being off-balance makes you vulnerable to grabs while you have a moderate amount of stamina damage (30 damage), forcing you to have to resist even passive grabs. This pairs _exceptionally_ well with tackling.

### Grappling

Grappling someone makes your unarmed attacks penetrate armor based on a new limb value called ``unarmed_effectiveness``. This is something shared by kicking.

### Unarmed Attacks in General

``unarmed_effectiveness`` has also taken over the functionality of ``unarmed_stun_threshold``, as well as accuracy calculations. Human equivalent limbs (pretty much all of them except mushrooms and golems) have a value of 10.

Now, ``unarmed_effectiveness`` determines how accurately a given limb makes unarmed attacks. Unarmed attacks have a base inaccuracy of 20%, with effectiveness acting as a reduction to this value. (so for humans, that's 20% - 10% before any value changes from brute and stamina damage). It is also capped at 75% miss chance, just to avoid those weird instances of two brawling fighters being incapable of finishing each other off at a certain amount of damage and it being real awkward, like it does currently.

It also determines the base probability of landing a knockdown punch. For humans, this is 10%. 

For the most part, these two particular changes are roughly equivalent to the current values, just handled in a way that is more straightforward to understand from a code perspective.

In addition to the above, human equivalent limbs have higher damage floors for unarmed attacks. Arms deal 5-10 damage, while legs deal 7-15 damage. In addition, kicks also deal stamina damage, like punches do.

### Minor Mentions

Golems and Mushroom People (who don't even use their limbs for their unarmed strikes because mushroom people start with a martial art) have very accurate punches, and their punches penetrate quite a bit of armor when they are entitled to that. They also have a high knockdown probability. This is partially because they previously already _had_ these features due to the wonky math at play, but also because this is their big thing they are good at.

Carp mutation also got a big win out of this as well. If and when you actually manage to get that to work and matter.

## Why It's Good For The Game

My favorite thing in this game is the robustness of unarmed fighting. It's the part of the game that actually acknowledges the sandbox and environmental interaction in a big way. The only problem with the unarmed combat is that it is a bit disjointed, and often much weaker than using even the most pathetic weapon you can get your hands on unless you're using the stun loops available. Those loops get a bit boring, even if they're mostly all environmental (except for the lucky neckgrab finish). Giving more options generally means that even when not in an ideal position, you still have _some_ options.

It also has some internal inconsistencies in design even in the same proc, like accuracy calculations and knockdowns, as well as weird splits in damage. So I decided to resolve that.

Now, every part of unarmed fighting has some relevance in the other parts. Predominantly, it is heavily favoured towards dealing stamina damage, making unarmed combat very favourable as a nonlethal method of taking someone down, which is something we currently lack considerably. While people may still opt to simply beat someone into actual crit rather than stop at stamina crit, the possibility is actually entirely available and supported now. No just banking on a lucky neckgrab after a shove.

Paying attention to damage dealt and thinking intelligently about how you apply combinations of effects allows even someone on the significant back foot an opportunity for a comeback if they know what they're doing against even armed opponents.

Separating accuracy and knockdown effectiveness from damage allows for more consistent design and readability, but also preventing weirdness ike tighter damage spreads increasing knockdown probabilities as well as increasing accuracy without the coder knowing why. This also lets us make unarmed attacks just that little bit stronger. Since unarmed attacks require more complicated combinations to work, I think this won't make them stronger than weapons necessarily, but it will make for more interesting swung fights.

## Changelog
:cl:
add: With the flood of Chi within the Spinward Sector receding, various masters of The Tunnel Arts, colloquially known as 'Maint-fu Masters', have started to refine the basics of their martial techniques. New forms have started to develop within Spacestation 13's hidden maintenance dojos.
add: Someone shoved off-balance makes them vulnerable to more guaranteed unarmed strikes, knockdowns from a successful punch, and more difficult to escape grabs.
add: Grabbing someone (as well as kicking them while they're on the floor) makes them more vulnerable to taking unarmed attack damage, even if they have armor.
balance: Unarmed strikes made with human-equivalent limbs have higher damage floors, meaning you overall do more damage on average while not increasing the overall damage potential. It's more consistent!
refactor: Significantly changed how punching accuracy and knockdowns are calculated.
balance: Golem and mushroom limbs are a lot more effective at punching as a result of these various changes. As they should be.
/:cl:

### CODE REVIEW

**Comment by moocowswag** [comment]
does this just make the shove shove cable cuff grab grab combo stronger?

**Comment by SinfulBliss** [comment]
Does this remove any current unarmed combat interactions (i.e., knockdowns and stuns) or does it merely add in inaccuracy from stam damage?

**Comment by mc-oofert** [comment]
ayo shovecuff buff

**Comment by necromanceranne** [comment]
> Does this remove any current unarmed combat interactions (i.e., knockdowns and stuns) or does it merely add in inaccuracy from stam damage?

No.

**Comment by necromanceranne** [comment]
> does this just make the shove shove cable cuff grab grab combo stronger? I feel like these changes impact that the most. (stronger grabs after shove)
 >
> you can easily get that 30 stamina from a single baton hit. (baton shove cable cuff grab grab?)
>
> Since security and heads of staff can consistently get that damage with a single baton hit or lethal energy gun hit uhhh i feel like this buffs them the most.

Since you'd need to do that amount of stamina damage to get an easier grab, and the combination you're talking about is all stuns and very much works already, I don't think it will meaningfully do much to make that stronger, as it is entirely relying on stun loops to function. It does make grabs immediately after a stun baton hit (specifically, security's stun baton) stronger, however, you're correct. But shove stunning is in of itself what is enabling this.

Specifically, stun looping after a baton hit is already possible and already strong due to floor shoves. There isn't anything here that should make that worse in much of a meaningful fashion in my view. Getting hit with a baton once is usually a defeat and will remain that way.

There is also no interactions in unarmed calculations that use burn damage, just brute damage. A lethal laser won't do you any good. Passive grabs also don't remove handuse, unlike aggressive grabs, so you still can protect yourself if you shove them off you when they get you into a grab. At that point, I'd argue just going for a baton hit would be a better use of your click interaction than banking on the passive grab resist RNG.

**Comment by DrDiasyl** [comment]
Unarmed combat must be much more interesting and badass then armed combat.
Yeah, winning against an enemy with a baton is rather cool, but winning against them in a fist fight is much cooler.

**Comment by nikothedude** [comment]
we are one step closer to a sekiro-style posture system. we can do it guys

**Comment by atlasle** [comment]
just a small suggestion, could you make it so you can visibly tell if someone is unbalanced, maybe like a very subtle little wobbling animation

**Comment by DrDiasyl** [comment]
> just a small suggestion, could you make it so you can visibly tell if someone is unbalanced, maybe like a very subtle little wobbling animation

Some of our emotes already do that, we can just force people to play them!
Although this way someone will try to deceive the opponent by intentionally playing them, but I think that's rather interesting.

**Comment by itseasytosee** [comment]
I had this an idea a lot like this while back. Here's an old shitty document if you are interested, there are still a few good ideas: https://hackmd.io/t_0WD8wgQUaeFxkhUvdhRQ

**Comment by necromanceranne** [comment]
I'll give it a read when I can!

**Comment by Likteer** [comment]
> > a subtle wobbling animation if someone is unbalanced
> 
> force people to play emotes that already do that

Shaking/twitching is already associated with shocks/stun weapons.
I suggest sped up/shortened waddle animation.

**Comment by necromanceranne** [comment]
I'll see what I can do.

**Comment by DrDiasyl** [comment]
> > > a subtle wobbling animation if someone is unbalanced
> > 
> > force people to play emotes that already do that
> 
> Shaking/twitching is already associated with shocks/stun weapons.
> I suggest sped up/shortened waddle animation.
> 
> 

Sway has animation, we can use it

**Comment by atlasle** [comment]
now when are we getting our command grabs and judo throws

**Comment by necromanceranne** [comment]
I've added the sway on the shove effects.

**Comment by necromanceranne** [comment]
Hm that's not good. I'll have to do this some other way.

**Comment by Jacquerel** [comment]
Is this PR actually ready? The last commit message makes it sound like a WIP.

File: code/modules/mob/living/carbon/human/_species.dm [hunk 1]
```diff
@@ -1160,16 +1160,25 @@ GLOBAL_LIST_EMPTY(features_by_species)
 				return FALSE
 		user.do_attack_animation(target, atk_effect)
 
+		var/off_balance = FALSE //has our target been shoved recently? If so, they're off-balance and we get an easy hit.
+		var/grappled = FALSE //Someone in a grapple is much more vulnerable to being harmed by punches.
+		if(target.has_movespeed_modifier(/datum/movespeed_modifier/shove))
+			off_balance = TRUE
+
+		if(target.pulledby && target.pulledby.grab_state >= GRAB_AGGRESSIVE)
+			grappled = TRUE
```
Comment (Jacquerel) [inline code/modules/mob/living/carbon/human/_species.dm:1169]:
```suggestion
		//has our target been shoved recently? If so, they're off-balance and we get an easy hit.
		var/off_balance = target.has_movespeed_modifier(/datum/movespeed_modifier/shove)
		//Someone in a grapple is much more vulnerable to being harmed by punches.
		var/grappled = target.pulledby && target.pulledby.grab_state >= GRAB_AGGRESSIVE
```
Not sure this is made more readable by being split across two lines?

File: code/modules/mob/living/carbon/human/_species.dm [hunk 2]
```diff
@@ -1195,23 +1204,25 @@ GLOBAL_LIST_EMPTY(features_by_species)
...
 			log_combat(user, target, "punched")
 
-		if((target.stat != DEAD) && damage >= attacking_bodypart.unarmed_stun_threshold)
+		//If we rolled a punch high enough to hit our stun threshold, or our target is off-balance and they have at least 40 damage+stamina loss, we knock them down
+		if((target.stat != DEAD) && prob(limb_accuracy) || (target.stat != DEAD) && off_balance && (target.getStaminaLoss() + user.getBruteLoss()) >= 40)
 			target.visible_message(span_danger("[user] knocks [target] down!"), \
 							span_userdanger("You're knocked down by [user]!"), span_hear("You hear aggressive shuffling followed by a loud thud!"), COMBAT_MESSAGE_RANGE, user)
 			to_chat(user, span_danger("You knock [target] down!"))
-			var/knockdown_duration = 40 + (target.getStaminaLoss() + (target.getBruteLoss()*0.5))*0.8 //50 total damage = 40 base stun + 40 stun modifier = 80 stun duration, which is the old base duration
+			var/knockdown_duration = 40 + (target.getStaminaLoss() + (target.getBruteLoss()*0.5))*0.8 //50 total damage = 40 base stun + 40 stun modifier = 8 second knockdown duration
```
Comment (Jacquerel) [inline code/modules/mob/living/carbon/human/_species.dm:1225]:
```suggestion
			var/knockdown_duration = 4 SECONDS + (target.getStaminaLoss() + (target.getBruteLoss()*0.5))*0.8 //50 total damage = 40 base stun + 40 stun modifier = 8 second knockdown duration
```
unless I am confusing what your comment means here

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by necromanceranne** [comment]
Aye, it's ready after I handle the reviews.

File: code/modules/mob/living/carbon/human/_species.dm (hunk 1 above)
Comment (necromanceranne) [reply code/modules/mob/living/carbon/human/_species.dm:1169]: Funnily enough I thought I already did this. My bad, I'll clean this up.

File: code/modules/mob/living/carbon/human/_species.dm (hunk 2 above)
Comment (necromanceranne) [reply code/modules/mob/living/carbon/human/_species.dm:1225]: Oh that's not my comment. And I also thought I had changed this too. I reckon I must have accidentally reverted a bit of cleanup at some point, I'll make that and the comment a lot more clearer.

**Comment by itseasytosee** [comment]
I hope this will come with an update to a wiki?

**Comment by necromanceranne** [comment]
Sure thing

**Pull request merged**